cql3/expr/prepare_expr.cc

```diff
@@ -870,8 +870,22 @@ try_prepare_expression(const expression& expr, data_dictionary::database db, con
         [] (const constant&) -> std::optional<expression> {
             on_internal_error(expr_logger, "Can't prepare constant_value, it should not appear in parser output");
         },
-        [&] (const binary_operator&) -> std::optional<expression> {
-            on_internal_error(expr_logger, "binary_operators are not yet reachable via prepare_expression()");
+        [&] (const binary_operator& binop) -> std::optional<expression> {
+            if (receiver.get() != nullptr && &receiver->type->without_reversed() != boolean_type.get()) {
```

Comment:
It's better to use `operator==(const abstract_type& x, const abstract_type& y)` without taking the address. This is safer if we want to stop interning types one day (internally it compares addresses).